feature(typescript_indexer): extend symbol cache to store contexts

ayazhafiz opened the pull request:
Create a proper caching instance for the TypeScript indexer that caches
VNames for a symbol under a particular TypeScript namespace and symbol
declaration context.

Cache lookups are done by querying a 2-layer map, mapping from a symbol
to a serialized namespace and context to a VName. The namespace and
context must be serialized because object instances are always unique in
JavaScript, and creating a new object for each lookup would render the
cache useless. An alternative to this would be to store an object with
namespace/context tuples:

```typescript
interface { [nsctx: [TSNamespace, Context]]: VName }
```

but TypeScript doesn't support this.

This PR also moves some blocks of code around so they make more sense in
their spatial and contextual locations.

ayazhafiz commented:
Let's get #3909 first before merging this.
cc @evmar, @zzmp

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -24,6 +24,20 @@ import * as utf8 from './utf8';
 const LANGUAGE = 'typescript';
 
+/**
+ * A indexer plugin adds extra functionality with the same inputs as the base
```
nit: An indexer plugin...

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -216,18 +146,154 @@ class StandardIndexerContext implements IndexerHost {
+enum Context {
+  /** No disambiguation about a node's declarations. */
+  Any,
+
```
Unnecessary blank lines.

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -216,18 +146,154 @@ class StandardIndexerContext implements IndexerHost {
+    return undefined;
+  }
+
+  /** Set a symbol VName for a given namespace and context. */
```
This always overrides set names. Is that desired?
I'd expect it to throw if the ts.Symbol and nsCtx were already in the map, as changing it might break other references.

ayazhafiz replied:
Fixed.

zzmp replied:
This could also be added to the docstring for `set`.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -216,18 +146,154 @@ class StandardIndexerContext implements IndexerHost {
+
+    // Set the symbol VName for the given namespace and `Any` context, if it has
+    // not already been set.
+    const nsAny = this.serialize(ns, Context.Any);
```
This is only necessary if `context != Context.Any`.
It might be more performant if you included that check, although I think the logic is simpler as-is.

ayazhafiz replied:
I think it's okay; it's definitely more readable this way.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -568,39 +634,32 @@ class Visitor {
+    // Disambiguate symbols with multiple declarations using a context.
+    if (sym.declarations.length > 1) {
       switch (context) {
+        // Getter and setter contexts are filtered for getter and setter
```
This seems self-explanatory - I don't think these comments are necessary.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -568,39 +634,32 @@ class Visitor {
+          break;
       }
     }
-    // Otherwise, if there are multiple declarations but no context is
```
Will deleting this codepath change behavior? Does it matter?
Now that you're defaulting to Context.Any, won't it change the behavior if there are multiple declarations, from using the get accessor to using whichever decl happens to be first?

ayazhafiz replied:
No, it doesn't change the behavior. This is an artifact of the work done to emit an implicit property on getters and setters:

https://github.com/kythe/kythe/blob/f8c4e9bb682adb3b2041e49d72029e6be3977e5f/kythe/typescript/indexer.ts#L1168-L1174

that we never removed.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -844,21 +898,28 @@ class Visitor {
+    // and mark the local symbol with the remote symbol's VName so that all
     // references resolve to the remote symbol.
-    this.symbolNames.set(localSym, [kImportType, kImportValue, kImportType]);
+    let kType, kValue;
```
Why kVarName if it's not a constant?

ayazhafiz replied:
Could you elaborate on your question?

zzmp replied:
This is more of a style question: why did you name the variable `kType, kValue` instead of `type, value`? I thought the `k` prefix was specified for constants.

ayazhafiz replied:
`k` is prefixed on variables that are VNames, seemingly standing for `kythe`.

evmar replied:
Your guess is right.  This naming convention of mine is dumb, I am sorry.  I'd be happy to replace it with something that makes more sense.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -844,21 +898,28 @@ class Visitor {
-        this.getSymbolName(remoteSym, TSNamespace.TYPE) :
-        null;
-    // Mark the local symbol with the remote symbol's VName so that all
+    // Attempt to create the Kythe VName for both the symbol's type and value
```
This changes behavior by not setting the Namespace. Does it matter?

zzmp replied:
?

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -844,21 +898,28 @@ class Visitor {
 
```
I thought you were moving this aliasing logic into the SymbolVNameCache. Are you still planning on that, or is this the final state?

zzmp replied:
?

ayazhafiz commented:
> This is hard to review because it mixes code moves with code changes, can you split it?

yep, that's my bad. Fixed. Also renamed the VName cache to `SymbolVNameStore` to better represent its purpose.

evmar commented on kythe/typescript/indexer.ts:
```diff
@@ -189,6 +191,78 @@ function todo(sourceRoot: string, node: ts.Node, message: string) {
+ *
+ * The `Any` context makes no guarantee of symbol declaration disambiguation.
+ * As a result, unless explicitly set for a given symbol and namespace, the
+ * VName of an `Any` context is lazily set to the VName of an arbitrary context.
```
Does "lazily" here mean that it will be sensitive to order of operations?

ayazhafiz replied:
yes. there is no guarantee of the assigned VName form.

evmar commented on kythe/typescript/indexer.ts:
```diff
@@ -189,6 +191,78 @@ function todo(sourceRoot: string, node: ts.Node, message: string) {
+ */
+type NamespaceAndContext = string&{__brand: 'nsctx'};
+class SymbolVNameStore {
+  private readonly cache =
```
Still not a cache, right?  If so, grep for more 'cache' in this change too.

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -189,6 +191,78 @@ function todo(sourceRoot: string, node: ts.Node, message: string) {
+ *              TYPE          Any
+ * ts.Symbol -> VALUE         Getter  -> VName
+ *              NAMESPACE     Setter
+ *                            ...
```
Are there other symbols besides the 3 contexts that we select on?

ayazhafiz replied:
Nope, not currently.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -141,13 +141,15 @@ export enum TSNamespace {
+ * disambiguating a node's declarations if it has multiple.
  */
 export enum Context {
+  /** No disambiguation about a node's declarations. */
```
This is lazy, but that's not mentioned here. Could you include `@see SymbolVNameStore` here?

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -189,6 +191,78 @@ function todo(sourceRoot: string, node: ts.Node, message: string) {
+ * As a result, unless explicitly set for a given symbol and namespace, the
+ * VName of an `Any` context is lazily set to the VName of an arbitrary context.
+ */
+type NamespaceAndContext = string&{__brand: 'nsctx'};
```
I'm afraid the jsdoc may be associated with this when it is meant for SymbolVNameStore. Could you move this type above the jsdoc comment?

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -189,6 +191,78 @@ function todo(sourceRoot: string, node: ts.Node, message: string) {
+    let vnameMap = this.store.get(symbol);
+    const nsCtx = this.serialize(ns, context);
+    if (vnameMap) {
+      // If there is already a map for the symbol, check that a VName is not
```
Unnecessary comment - repeats the docstring.

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -189,6 +191,78 @@ function todo(sourceRoot: string, node: ts.Node, message: string) {
+      }
+      vnameMap.set(nsCtx, vname);
+    } else {
+      // If there is no map for the symbol, create a new one with an entry of
```
Unnecessary comment - repeats the docstring.

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -207,13 +281,9 @@ class StandardIndexerContext implements IndexerHost {
-   * The value is a tuple of the separate TypeScript namespaces, and entries
-   * in it correspond to TSNamespace values.  See the documentation of
-   * TSNamespace.
+   * symbolNames is a store of ts.Symbols to their assigned VNames.
```
Single-line this docstring (or remove it entirely).

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -496,39 +565,29 @@ class StandardIndexerContext implements IndexerHost {
-      if (getDecls.length > 0) declarations = getDecls;
-    }
 
     const decl = declarations[0];
```
Maybe comment that you are taking the first decl, and so if the Context is Any, it will be "lazily" decided.
Then again, that's been mentioned in at least two other places, so this is unnecessary.

ayazhafiz replied:
I think it's okay to have the documentation where it already is. Not adding here unless big opposition.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -541,33 +600,35 @@ class StandardIndexerContext implements IndexerHost {
-        null;
-    // Mark the local symbol with the remote symbol's VName so that all
+    // Attempt to create the Kythe VName for both the symbol's type and value
+    // and mark the local symbol with the remote symbol's VName so that all
```
s/VName/VName(s)

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -541,33 +600,35 @@ class StandardIndexerContext implements IndexerHost {
-        this.getSymbolName(remoteSym, TSNamespace.TYPE) :
-        null;
-    // Mark the local symbol with the remote symbol's VName so that all
+    // Attempt to create the Kythe VName for both the symbol's type and value
```
s/the/a or s/VName/VNames

ayazhafiz replied:
Done.

zzmp commented on kythe/typescript/indexer.ts:
```diff
@@ -607,8 +604,8 @@ class StandardIndexerContext implements IndexerHost {
-    // Attempt to create the Kythe VName for both the symbol's type and value
-    // and mark the local symbol with the remote symbol's VName so that all
+    // Attempt to create VNames for both the symbol's type and value and mark
+    // the local symbos with the remote symbol's VName(s) so that all
```
typo: symbos

ayazhafiz replied:
Done.
